com.hazelcast.client.ClientReconnectTest.testCallbackAfterClientShutdown

_5.1.z_ (commit 7e487db1061d71702261505ed968974903a7dcf2)

Failed on CorrettoJDK8: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-5.maintenance-CorrettoJDK8/41/testReport/com.hazelcast.client/ClientReconnectTest/testCallbackAfterClientShutdown/

<details><summary>Stacktrace:</summary>

```
com.hazelcast.client.HazelcastClientNotActiveException: Client is not active.
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.serializerFor(AbstractSerializationService.java:545)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:227)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:214)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:199)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:154)
	at com.hazelcast.client.impl.spi.ClientProxy.toData(ClientProxy.java:76)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.putAsyncInternal(ClientMapProxy.java:402)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.putAsync(ClientMapProxy.java:375)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.putAsync(ClientMapProxy.java:193)
	at com.hazelcast.client.ClientReconnectTest.testCallbackAfterClientShutdown(ClientReconnectTest.java:142)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

</details>

<details><summary>Standard output:</summary>

```
04:39:16,394 TRACE |testRequestShouldFailOnShutdown| - [ClientListenerService] hz.client_4140.eventRegistration- - hz.client_4140 [dev] [5.1.3-SNAPSHOT] Registered ClientListenerRegistration{codec=com.hazelcast.client.impl.spi.impl.ClientInvocationServiceImpl$1@6334aa89, handler=com.hazelcast.client.impl.spi.impl.ClientInvocationServiceImpl$BackupEventHandler@7ea4c735} to MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=true, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2022-07-09 04:39:16.393, lastWriteTime=2022-07-09 04:39:16.386, closedTime=never, connected server version=null}}
04:39:16,395  INFO |testReconnectToNewInstanceAtSameAddress| - [JetServiceBackend] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Setting number of cooperative threads and default parallelism to 8
04:39:16,395  INFO |testRequestShouldFailOnShutdown| - [ClientStatisticsService] testRequestShouldFailOnShutdown - Client statistics is enabled with period 5 seconds.
04:39:16,402  INFO |testExceptionAfterClientShutdown_fromClientConnectionManager| - [LifecycleService] testExceptionAfterClientShutdown_fromClientConnectionManager - hz.client_4139 [dev] [5.1.3-SNAPSHOT] HazelcastClient 5.1.3-SNAPSHOT (20220709 - 7e487db) is SHUTDOWN
04:39:16,404  INFO |testRequestShouldFailOnShutdown| - [PartitionStateManager] hz.distracted_snyder.partition-operation.thread-0 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Initializing cluster partition table arrangement...
04:39:16,405  INFO |testExceptionAfterClientShutdown_fromClientConnectionManager| - [LifecycleService] Thread-31531 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
04:39:16,405  WARN |testExceptionAfterClientShutdown_fromClientConnectionManager| - [Node] Thread-31531 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Terminating forcefully...
04:39:16,405  INFO |testExceptionAfterClientShutdown_fromClientConnectionManager| - [Node] Thread-31531 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Shutting down connection manager...
04:39:16,405  INFO |testExceptionAfterClientShutdown_fromClientConnectionManager| - [Node] Thread-31531 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Shutting down node engine...
04:39:16,406  INFO |testReconnectToNewInstanceAtSameAddress| - [Diagnostics] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
04:39:16,406  INFO |testReconnectToNewInstanceAtSameAddress| - [LifecycleService] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] [127.0.0.1]:5701 is STARTING
04:39:16,406  INFO |testReconnectToNewInstanceAtSameAddress| - [ClusterService] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5701 - 78afc254-95cd-490d-886c-7ba310474588 this
]

04:39:16,406  INFO |testRequestShouldFailOnShutdown| - [LifecycleService] testRequestShouldFailOnShutdown - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
04:39:16,406  INFO |testReconnectToNewInstanceAtSameAddress| - [JobCoordinationService] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Jet started scanning for jobs
04:39:16,406  INFO |testReconnectToNewInstanceAtSameAddress| - [LifecycleService] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] [127.0.0.1]:5701 is STARTED
04:39:16,406 DEBUG |testRequestShouldFailOnShutdown| - [JobCoordinationService] testRequestShouldFailOnShutdown - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Added a shutting-down member: 56b8fec8-c22c-4a29-90b6-a9ceb9f71036
04:39:16,406 DEBUG |testReconnectToNewInstanceAtSameAddress| - [JobCoordinationService] hz.ecstatic_snyder.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
04:39:16,407  INFO |testRequestShouldFailOnShutdown| - [MigrationManager] hz.distracted_snyder.cached.thread-17 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Shutdown request of Member [127.0.0.1]:5701 - 56b8fec8-c22c-4a29-90b6-a9ceb9f71036 this is handled
04:39:16,407  INFO |testRequestShouldFailOnShutdown| - [Node] testRequestShouldFailOnShutdown - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Shutting down connection manager...
04:39:16,407  WARN |testRequestShouldFailOnShutdown| - [TestClientRegistry$MockedServerConnection] testRequestShouldFailOnShutdown - Server connection closed: null
04:39:16,407  INFO |testRequestShouldFailOnShutdown| - [MockServer] testRequestShouldFailOnShutdown - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Removed connection to endpoint: [address=[127.0.0.1]:40001, uuid=16cef92d-0d5b-48c8-9767-f45329cc7bdb], connection: MockedNodeConnection{ remoteAddress = [127.0.0.1]:40001, localAddress = [127.0.0.1]:5701, connectionId = 1}
04:39:16,407  INFO |testRequestShouldFailOnShutdown| - [Node] testRequestShouldFailOnShutdown - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Shutting down node engine...
04:39:16,407  WARN |testRequestShouldFailOnShutdown| - [TcpClientConnection] pool-12437-thread-1 - hz.client_4140 [dev] [5.1.3-SNAPSHOT] MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=false, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2022-07-09 04:39:16.406, lastWriteTime=2022-07-09 04:39:16.406, closedTime=2022-07-09 04:39:16.407, connected server version=null}} closed. Reason: com.hazelcast.spi.exception.TargetDisconnectedException[Mocked Remote socket closed]
com.hazelcast.spi.exception.TargetDisconnectedException: Mocked Remote socket closed
	at com.hazelcast.client.test.TestClientRegistry$MockedTcpClientConnection$4.run(TestClientRegistry.java:332) [test-classes/:?]
	at com.hazelcast.client.test.TwoWayBlockableExecutor$BlockableRunnable.run(TwoWayBlockableExecutor.java:98) [test-classes/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_312]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_312]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_312]
04:39:16,407  WARN |testRequestShouldFailOnShutdown| - [TwoWayBlockableExecutor] pool-12437-thread-1 - Dropping outgoing runnable since other end closed. Client Closed EOF. MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=false, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2022-07-09 04:39:16.406, lastWriteTime=2022-07-09 04:39:16.406, closedTime=2022-07-09 04:39:16.407, connected server version=null}}
04:39:16,407  INFO |testRequestShouldFailOnShutdown| - [ClientConnectionManager] pool-12437-thread-1 - hz.client_4140 [dev] [5.1.3-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5701:56b8fec8-c22c-4a29-90b6-a9ceb9f71036, connection: MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=false, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2022-07-09 04:39:16.406, lastWriteTime=2022-07-09 04:39:16.406, closedTime=2022-07-09 04:39:16.407, connected server version=null}}
04:39:16,407  INFO |testRequestShouldFailOnShutdown| - [LifecycleService] pool-12437-thread-1 - hz.client_4140 [dev] [5.1.3-SNAPSHOT] HazelcastClient 5.1.3-SNAPSHOT (20220709 - 7e487db) is CLIENT_DISCONNECTED
04:39:16,407  INFO |testRequestShouldFailOnShutdown| - [ClientConnectionManager] hz.client_4140.internal-1 - hz.client_4140 [dev] [5.1.3-SNAPSHOT] Trying to connect to cluster: dev
04:39:16,407  INFO |testRequestShouldFailOnShutdown| - [ClientConnectionManager] hz.client_4140.internal-1 - hz.client_4140 [dev] [5.1.3-SNAPSHOT] Trying to connect to Member [127.0.0.1]:5701 - 56b8fec8-c22c-4a29-90b6-a9ceb9f71036
04:39:16,407  WARN |testRequestShouldFailOnShutdown| - [ClientConnectionManager] hz.client_4140.internal-1 - hz.client_4140 [dev] [5.1.3-SNAPSHOT] Exception during initial connection to Member [127.0.0.1]:5701 - 56b8fec8-c22c-4a29-90b6-a9ceb9f71036: com.hazelcast.core.HazelcastException: java.io.IOException: Can not connect to [127.0.0.1]:5701: instance does not exist
04:39:16,407  WARN |testRequestShouldFailOnShutdown| - [ClientConnectionManager] hz.client_4140.internal-1 - hz.client_4140 [dev] [5.1.3-SNAPSHOT] Unable to get live cluster connection, retry in 1000 ms, attempt: 1, cluster connect timeout: 2000 ms, max backoff: 30000 ms
04:39:16,436  INFO |testCallbackAfterClientShutdown| - [LifecycleService] Thread-31532 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
04:39:16,437  INFO |testCallbackAfterClientShutdown| - [Node] Thread-31532 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Node is already shutting down... Waiting for shutdown process to complete...
04:39:16,437  INFO |testCallbackAfterClientShutdown| - [LifecycleService] Thread-31532 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after testCallbackAfterClientShutdown(com.hazelcast.client.ClientReconnectTest): BuildInfo{version='5.1.3-SNAPSHOT', build='20220709', buildNumber=20220709, revision=7e487db, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'testCallbackAfterClientShutdown(com.hazelcast.client.ClientReconnectTest):'
04:39:10, accumulated pauses: 2142 ms, max pause: 2097 ms, pauses over 1000 ms: 1
04:39:15, accumulated pauses: 230 ms, max pause: 229 ms, pauses over 1000 ms: 0

04:39:16,439  INFO |testClientShutdownIfReconnectionNotPossible| - [NodeExtension] testClientShutdownIfReconnectionNotPossible - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Destroying node NodeExtension.
04:39:16,440  INFO |testClientShutdownIfReconnectionNotPossible| - [Node] testClientShutdownIfReconnectionNotPossible - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Hazelcast Shutdown is completed in 2420 ms.
04:39:16,440  INFO |testClientShutdownIfReconnectionNotPossible| - [LifecycleService] testClientShutdownIfReconnectionNotPossible - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN

No metrics recorded during the test
04:39:16,444  INFO |testExceptionAfterClientShutdown_fromClientConnectionManager| - [NodeExtension] Thread-31531 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Destroying node NodeExtension.
04:39:16,444  INFO |testExceptionAfterClientShutdown_fromClientConnectionManager| - [Node] Thread-31531 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Hazelcast Shutdown is completed in 39 ms.
04:39:16,444  INFO |testExceptionAfterClientShutdown_fromClientConnectionManager| - [LifecycleService] Thread-31531 - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
04:39:16,446  INFO |testRequestShouldFailOnShutdown| - [NodeExtension] testRequestShouldFailOnShutdown - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Destroying node NodeExtension.
04:39:16,446  INFO |testRequestShouldFailOnShutdown| - [Node] testRequestShouldFailOnShutdown - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] Hazelcast Shutdown is completed in 40 ms.
04:39:16,446  INFO |testRequestShouldFailOnShutdown| - [LifecycleService] testRequestShouldFailOnShutdown - [127.0.0.1]:5701 [dev] [5.1.3-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
```

</details>

Failed on `master` on OracleJDK8: https://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-OracleJDK8/208/testReport/com.hazelcast.client/ClientReconnectTest/testCallbackAfterClientShutdown/

<details><summary>Stacktrace is slightly different:</summary>

```
java.lang.AssertionError: com.hazelcast.spi.exception.TargetDisconnectedException: Mocked Remote socket closed is not an instanceof com.hazelcast.client.HazelcastClientNotActiveException
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at com.hazelcast.test.HazelcastTestSupport.assertInstanceOf(HazelcastTestSupport.java:930)
	at com.hazelcast.client.ClientReconnectTest.testCallbackAfterClientShutdown(ClientReconnectTest.java:151)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

</details>

<details><summary>Standard output:</summary>

```
22:42:57,016  INFO |testClientReconnectOnClusterDown| - [Diagnostics] testClientReconnectOnClusterDown - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
22:42:57,017  INFO |testClientReconnectOnClusterDown| - [LifecycleService] testClientReconnectOnClusterDown - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5702 is STARTING
22:42:57,017  WARN |testReconnectToNewInstanceAtSameAddress| - [CPSubsystem] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
22:42:57,017  INFO |testCallbackAfterClientShutdown| - [LifecycleService] Thread-32550 - hz.client_4113 [dev] [5.2-SNAPSHOT] HazelcastClient 5.2-SNAPSHOT (20220724 - a12b871) is SHUTDOWN
22:42:57,017  INFO |testClientReconnectOnClusterDown| - [ClusterService] testClientReconnectOnClusterDown - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5702 - b9fdac83-3016-4804-ae86-026c1519162f this
]

22:42:57,017  INFO |testCallbackAfterClientShutdown| - [LifecycleService] Thread-32550 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
22:42:57,017  INFO |testCallbackAfterClientShutdown| - [Node] Thread-32550 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Node is already shutting down... Waiting for shutdown process to complete...
22:42:57,017  INFO |testCallbackAfterClientShutdown| - [LifecycleService] Thread-32550 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after testCallbackAfterClientShutdown(com.hazelcast.client.ClientReconnectTest): BuildInfo{version='5.2-SNAPSHOT', build='20220724', buildNumber=20220724, revision=a12b871, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'testCallbackAfterClientShutdown(com.hazelcast.client.ClientReconnectTest):'
22:42:55, accumulated pauses: 1322 ms, max pause: 985 ms, pauses over 1000 ms: 0

22:42:57,017  WARN |testClientReconnectOnClusterDown| - [Node] testClientReconnectOnClusterDown - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Config seed port is 5701 and cluster size is 1. Some of the ports seem occupied!
22:42:57,017  INFO |testClientReconnectOnClusterDown| - [JobCoordinationService] testClientReconnectOnClusterDown - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Jet started scanning for jobs
22:42:57,017  INFO |testClientReconnectOnClusterDown| - [LifecycleService] testClientReconnectOnClusterDown - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5702 is STARTED
22:42:57,017 DEBUG |testClientReconnectOnClusterDown| - [JobCoordinationService] hz.thirsty_carver.cached.thread-1 - [127.0.0.1]:5702 [dev] [5.2-SNAPSHOT] Not starting jobs because partitions are not yet initialized.

No metrics recorded during the test
22:42:57,019  INFO |testReconnectToNewInstanceAtSameAddress| - [JetServiceBackend] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Setting number of cooperative threads and default parallelism to 8
22:42:57,020  INFO |testReconnectToNewInstanceAtSameAddress| - [Diagnostics] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
22:42:57,020  INFO |testReconnectToNewInstanceAtSameAddress| - [LifecycleService] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is STARTING
22:42:57,020  INFO |testReconnectToNewInstanceAtSameAddress| - [ClusterService] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5701 - f3c398fc-ed33-4808-a493-bd27c01ba46f this
]

22:42:57,020  INFO |testReconnectToNewInstanceAtSameAddress| - [JobCoordinationService] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Jet started scanning for jobs
22:42:57,020 DEBUG |testReconnectToNewInstanceAtSameAddress| - [JobCoordinationService] hz.agitated_carver.cached.thread-2 - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
22:42:57,020  INFO |testReconnectToNewInstanceAtSameAddress| - [LifecycleService] testReconnectToNewInstanceAtSameAddress - [127.0.0.1]:5701 [dev] [5.2-SNAPSHOT] [127.0.0.1]:5701 is STARTED
```

</details>

--- a/hazelcast/src/test/java/com/hazelcast/client/ClientReconnectTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/client/ClientReconnectTest.java
@@ -27,7 +27,6 @@ import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.LifecycleEvent;
 import com.hazelcast.map.IMap;
 import com.hazelcast.test.HazelcastParallelClassRunner;
-import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.ParallelJVMTest;
 import com.hazelcast.test.annotation.QuickTest;
 import org.junit.After;
@@ -36,6 +35,7 @@ import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
 import java.util.Iterator;
+import java.util.UUID;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicReference;
@@ -45,7 +45,7 @@ import static org.junit.Assert.assertNull;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelJVMTest.class})
-public class ClientReconnectTest extends HazelcastTestSupport {
+public class ClientReconnectTest extends ClientTestSupport {
 
     private final TestHazelcastFactory hazelcastFactory = new TestHazelcastFactory();
 
@@ -60,7 +60,7 @@ public class ClientReconnectTest extends HazelcastTestSupport {
         ClientConfig clientConfig = new ClientConfig();
         clientConfig.getConnectionStrategyConfig().getConnectionRetryConfig().setClusterConnectTimeoutMillis(Long.MAX_VALUE);
         final HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
-        ClientTestSupport.ReconnectListener reconnectListener = new ClientTestSupport.ReconnectListener();
+        ReconnectListener reconnectListener = new ReconnectListener();
         client.getLifecycleService().addLifecycleListener(reconnectListener);
         IMap<String, String> m = client.getMap("default");
         h1.shutdown();
@@ -131,14 +131,20 @@ public class ClientReconnectTest extends HazelcastTestSupport {
     }
 
     @Test
-    public void testCallbackAfterClientShutdown() {
+    public void testCallbackAfterServerShutdown() {
         final HazelcastInstance server = hazelcastFactory.newHazelcastInstance();
         ClientConfig clientConfig = new ClientConfig();
         clientConfig.getConnectionStrategyConfig().getConnectionRetryConfig().setClusterConnectTimeoutMillis(2000);
         HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
 
         IMap<Object, Object> test = client.getMap("test");
+        UUID serverUuid = server.getLocalEndpoint().getUuid();
         server.shutdown();
+
+        // wait until client disconnects from the shutdown member to not get
+        // target disconnected exception in case the putAsync call is made
+        // before the connection on the client side is closed.
+        makeSureDisconnectedFromServer(client, serverUuid);
         CompletionStage<Object> future = test.putAsync("key", "value");
         CountDownLatch latch = new CountDownLatch(1);
         AtomicReference<Throwable> reference = new AtomicReference<>();

